src/FSharpVSPowerTools.Core/Utils.fs

```diff
@@ -27,60 +48,136 @@ module List =
             | _ -> List.rev acc
         loop [] xs
 
+    /// Fold over the list passing the index and element at that index to a folding function
+    let foldi (folder:'State -> int -> 'T -> 'State) (state:'State) (xs:'T list) =        
+        match xs with 
+        | [] -> state
+        | _ -> 
+            let f = OptimizedClosures.FSharpFunc<_,_,_,_>.Adapt folder
+            let rec loop idx s xs = 
+                match xs with 
+                | [] -> s
+                | h::t -> loop (idx+1) (f.Invoke(s,idx,h)) t
+            loop 0 state xs
+
+
+    let inline groupBy (keyfn:'T->'Key) (list: 'T list) =
+        let dict = Dictionary<'Key,ResizeArray<'T>> (HashIdentity.Structural)
+        // Build the groupings
+        let rec loop list =
+            match list with
+            | v :: t -> 
+                let key = keyfn v
+                let ok,prev = dict.TryGetValue key
+                if ok then
+                    prev.Add v
+                else 
+                    let prev = new ResizeArray<'T> 1
+                    dict.[key] <- prev
+                    prev.Add v
+                loop t
+            | _ -> ()
+        loop list
+        dict  // Return the list-of-lists.
+        |> Seq.map (fun group -> (group.Key, group.Value.ToArray() |> List.ofArray))
```

Comment:
that was a bad key function
with `let inline keyfn (str:string) = str.Length % 4`
Seq is better 
```
groupby seq 0.13ms
groupby arr 0.41ms
Ran 100 times
Ratio:  0.3170731707
groupby seq 1.77ms
groupby arr 1.89ms
Ran 100 times
Ratio:  0.9365079365
groupby seq 21.36ms
groupby arr 18.18ms
Ran 100 times
Ratio:  1.174917492
groupby seq 203.43ms
groupby arr 202.44ms
Ran 100 times
Ratio:  1.004890338
```